### PR TITLE
Add Cgroup V1 block IO integration test

### PR DESCRIPTION
**Pull request by YJDoc2**
This adds Cgroups V1 block io integration test

### CODE REVIEW

**Comment by YJDoc2** [comment]
The CI is likely to fail for the same reason as https://github.com/containers/youki/pull/528#issuecomment-991806762
The test also give same error for youki on my local machine, but runc passes all the tests.
Once that issue is fixed, I'll run the tests on youki as well and update here.

**Comment by Furisto** [comment]
@YJDoc2 I fixed this problem [here](https://github.com/containers/youki/pull/528/commits/d1bdf10859cb0077ba828ee0318e4212b29e78b2)

**Comment by YJDoc2** [comment]
Oh yes, I tested youki with that change and it passes the tests as well. I'm waiting for that PR to be merged so it will pass in the CI itself, and the I can ask for reviews. Thanks :+1:

File: crates/integration_test/src/tests/cgroups/blkio.rs [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+            fs::read_to_string(format!("{}/blkio.throttle.write_bps_device", path))
+                .context("error in reading block io write bps device")?;
+        let mut throttle_devices = Vec::new();
+        for line in throttle_write_bps_string.lines() {
+            let (device_data, rate) = line
+                .split_once(" ")
+                .context("invalid weight device format")?;
+            let (major_str, minor_str) = device_data
+                .split_once(":")
+                .context("invalid major-minor number format")?;
```
Comment (Furisto) [inline crates/integration_test/src/tests/cgroups/blkio.rs:215]: This snippet seems to be used a few times. Could be worth creating a method for this.

File: crates/integration_test/src/tests/cgroups/blkio.rs [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+        .join("blkio/runtime-test")
+        .join(cgroup_name);
+    let block_io = get_blkio_data(cgroup_path.to_str().unwrap())?;
+
+    let resources = spec.linux().as_ref().unwrap().resources().as_ref().unwrap();
+    let spec_block_io = resources.block_io().as_ref().unwrap();
+    // weight ------
+    if supports_weight() {
+        if spec_block_io.weight().is_none() {
+            return Err(anyhow::anyhow!("spec block io weight is none"));
```
Comment (Furisto) [inline crates/integration_test/src/tests/cgroups/blkio.rs:292]:
```suggestion
            bail!("spec block io weight is none");
```
A few more instances of this.

**Comment by YJDoc2** [comment]
Hey, sorry for the mistakes :sweat_smile: :sweat_smile:
I had realized that the device reading code can be extracted into function, but forgot to do that before making commit and asking review :sweat_smile: 

I have fixed that, as well as improved error reporting as you have suggested. One of the issue that it caused is that we need to clone the path, as fs::read_to_string takes the path by value. 

I'm not sure why read_to_string needs to take it by value, as I checked rust src, and it eventually calls as_ref on it, which needs only by ref, not by value... :thinking: 

Anyways, please take a look now, thanks :)

**Comment by Furisto** [comment]
> I have fixed that, as well as improved error reporting as you have suggested. One of the issue that it caused is that we need to clone the path, as fs::read_to_string takes the path by value.
> 
Can you point to a specific example? PathBuf implements AsRef<Path> so you should not need to clone it.

**Comment by YJDoc2** [comment]
> Can you point to a specific example? PathBuf implements AsRef so you should not need to clone it.

Yes you're right, I have been using it incorrectly :sweat: 

I have removed the clone calls, so PTAL.

**Comment by Furisto** [comment]
Looks good! @utam0k Will this conflict with #514? If it does it might make more sense to merge yours first, as yours looks trickier to adapt to the changes than the other way around.

**Comment by utam0k** [comment]
> Looks good! @utam0k Will this conflict with #514? If it does it might make more sense to merge yours first, as yours looks trickier to adapt to the changes than the other way around.

Thank you for taking care of my pr. I still have a few things to do on that PR of mine, so you can merge this one first!

**Review by Furisto — APPROVED** [review]

**Pull request merged**